compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt

```diff
@@ -504,6 +533,9 @@ class ComposeScene internal constructor(
         // - move from one point of the window to another (owner == lastMoveOwner): Move
         // - move from one popup to another (owner != lastMoveOwner): [Popup 1] Exit, [Popup 2] Enter
 
+        // TODO(demin) How it should behave for touch?
+        //  We need to decide whether we need Enter/Exit events
```

Comment:
Just to understand why this question even arise:
My understanding is that Move events with Touch always imply that the pointer state `isPressed=true` and it means that lastMoveOwner doesn't change even if a pointer goes out of owner's bounds. Is it right?

It seems to be the same as when I pressDown the mouseButton on an item, and then move the mouse out of the item bounds to another item while holding the mouse button pressed. The lastMoveOwner won't change in this case - it will be the same first item, right? And the second item (where I moved the mouse to) won't receive `Enter` event.

If the above is correct, then I think there is no need to do anything extra for Touch.

Reply:
> My understanding is that Move events with Touch always imply that the pointer state isPressed=true and it means that lastMoveOwner doesn't change even if a pointer goes out of owner's bounds. Is it right?

Yes, we still receive move events if the mouse/touch is pressed.

But this is not about keeping the owner on the pressed state. It is about "hover" events Enter/Exit, which were added additionally after we already had Move events. I am not sure how we should emit them in case of Touch. With mouse everything is simple - they are emitted when we hover the component. With touch we can't hover the component until we touch the screen.

Also, native Android behaves strange when we use Mouse - Enter/Exit emit correctly, but when we touch the screen, mouse disappears, and we receive an Exit event.

Reply:
> With touch we can't hover the component until we touch the screen.
yes. 

Do we need enter/exit events only for Hover? Are there other use cases? 
____________


This is not about this PR. Just some thoughts:

Let's consider a scenario:

There is a Box. Using Touch, I press on it and hold it pressed and Move the pointer within its bounds. 
Should it receive an  Enter event? It feels like it should not. Maybe because I think about Enter event as when a pointer is initially outside the box and then moves towards it and crosses the bounds of the box.

Another scenario:
There are Box1 and Box2 (without space between them). Using Touch, I press on Box1 and move the pointer into Box2. I think Box2 won't receive any event at all, since all events will be dispatched to Box1. Should Box1 receive Exit event when pointer moved out of its bounds?  If it didn't receive an Enter event, then it seems that receiving an Exit could weird.

Reply:
> Do we need enter/exit events only for Hover? Are there other use cases?

We could also play some sound signal on enter, or show a tooltip. But that all relate to hover. Maybe we can even tell that hover and Enter/Exit are the same things :). The question, does it make sense to treat touch inside a Box as Hover (or as Enter).

> event as when a pointer is initially outside the box

If the pointer is a finger (touch) then it is initially outside the box before the touch, and probably is inside during the touch.

> I think about Enter event as when a pointer is initially outside the box and then moves towards it and crosses the bounds of the box.

The pointer can also be initially inside the Box on the first appear, so there is no crossing of the bounds. If we treat Enter/Exit as the change of the `isPointerInsideBoxOrCaptured` state, it will simplify things.

> Box2 won't receive any event at all, since all events will be dispatched to Box1

We can send Enter as soon as Box1 loses the captured state (on a Release).

> If it didn't receive an Enter event (according to Scenario 1) , then it seems that receiving an Exit could be weird.

Yes, the general rule is we only receive an Exit, if there was an Enter. Otherwise there will be a lot of inconsistency.

After our conversation, I start to think that we should send Enter even on touch. Probably before the Press. And send Exit after the Release. It isn't the case now, because we only send Enter on Move. And it probably means that we need to create synthetic events (there was one physical press, but we will send Enter + Press).